### ISSUE TITLE
edgetpu_c.h should only #include C headers from TfLite

### ISSUE THREAD

**Issue by llschloesser** [google-coral/edgetpu#179]
edgetpu_c.h has [`#include "tensorflow/lite/context.h"`](https://github.com/google-coral/edgetpu/blob/7064b94dd5b996189242320359dbab8b52c94a84/libedgetpu/edgetpu_c.h#L58). This breaks the idea of a C API. This should instead be `#include "tensorflow/lite/c/c_api.h"` or similar as the only struct it needs is `TfLiteDelegate`. Am I missing something?

**Comment by Namburger** [comment]
@llschloesser 
That is true, but the all that context does is [#include "tensorflow/lite/c/common.h"](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/context.h#L19) which is a c api. Do you actually have any issue using a c compiler when building your application with our api?

**Comment by llschloesser** [comment]
I get an error because I don't have that header installed on the paths I am searching. I am compiling and linking against the C API and libtensorflowlite_c.so and not the C++ API and libtensorflow-lite.a. Making the change I suggested allows for a successful build.

**Comment by Namburger** [comment]
@llschloesser I see, let me discuss this with our team, thanks for reporting!

**Comment by llschloesser** [comment]
I should clarify. I get an error in my application build, not building libedgetpu. Regardless, I think including the C header is the better choice. Thanks for looking into it.

**Comment by llschloesser** [comment]
I am able to run a .tflite model (mobilenet_v2_1.0_224_quant.tflite) with the changes I proposed. However, after compiling to edgetpu model (mobilenet_v2_1.0_224_quant_edgetpu.tflite) I get the following errors:
```
ERROR: Encountered unresolved custom op: edgetpu-custom-op.
ERROR: Node number 0 (edgetpu-custom-op) failed to prepare.
```
```
$ dpkg -l | grep edgetpu
ii  edgetpu-compiler                                            14.1                                                      amd64        Edge TPU model compiler
ii  libedgetpu-dev                                              14.1                                                      amd64        Development files for libedgetpu
ii  libedgetpu1-max:amd64                                       14.1                                                      amd64        Support library for Edge TPU
ii  python3-edgetpu                                             14.1                                                      amd64        Edge TPU Python API
```
My `libtensorflowlite_c.so` version is 2.2.0.

Where is the C API to add the custom ops resolver?

**Comment by Namburger** [comment]
@llschloesser oh, you'll need to compile `libtensorflowlite_c.so` from this [tensorflow commit](https://github.com/google-coral/edgetpu/blob/master/WORKSPACE#L5) to be compatible with our current released library

**Comment by llschloesser** [comment]
@Namburger, after building the sha (some version of 2.1.X) of libtensorflowlite_c.so you pointed me to I am still seeing the same errors when trying run the edgetpu compiled .tflite model. `edgetpu_c.h` pointing at `tensorflow/lite/context.h` would make sense if I need to add the custom ops resolver which requires the C++ API. I think the real solution is to expose the custom ops resolver in the TfLite C API. I was hoping to use the C API as to my eyes it is much cleaner and easier to read than the C++ API.

**Comment by Namburger** [comment]
@llschloesser oh sorry, this could also occurs if you haven't registered the edgetpu ops, which I assumed you have. May I see how you initialized the interpreter in the c api?
it looks something like this for c++:
```
  auto* delegate =
      edgetpu_create_delegate(device.type, device.path, nullptr, 0);
  interpreter_->ModifyGraphWithDelegate({delegate, edgetpu_free_delegate});
```

**Comment by llschloesser** [comment]
In my header I have:
```
  ...
  TfLiteModel* model_;                                                          
  TfLiteInterpreter* interpreter_;                                              
  TfLiteInterpreterOptions* options_;                                           
  TfLiteDelegate* delegate_;
  ...
```

In my source file I have:
```
  ...
  model_ = TfLiteModelCreateFromFile(params_.model.c_str());
  options_ = TfLiteInterpreterOptionsCreate();
                                                       
  interpreter_ = TfLiteInterpreterCreate(model_, options_);                         
                                                                                   
  delegate_ = edgetpu_create_delegate(device.type, device.path, nullptr, 0);
  TfLiteInterpreterOptionsAddDelegate(options_, delegate_);
  ...
```

**Comment by llschloesser** [comment]
I got it working. It comes down to needing `edgetpu::kCustomOp` and `edgetpu::RegisterCustomOp()` which are defined in `edgetpu.h`. Which is fine, but one would think that `edgetpu_c.h` is complete, right?
```
  //
  // Register Edge TPU custom ops:
  //
  TfLiteInterpreterOptionsAddCustomOp(options_,
                                      edgetpu::kCustomOp,
                                      edgetpu::RegisterCustomOp(), 10, 13);
```

**Comment by llschloesser** [comment]
Turns out it was as simple as rearranging the call order, creating the interpreter after adding the delegate to the options, which I guess makes sense. Original issue is still valid, might as well just point at the C headers.

**Comment by Namburger** [comment]
@llschloesser a feature request is in place :)

**Comment by Namburger** [comment]
@llschloesser wait, can you clarify, in the end this is all that should be needed to create the interpreter right?
```
  delegate_ = edgetpu_create_delegate(device.type, device.path, nullptr, 0);
  TfLiteInterpreterOptionsAddDelegate(options_, delegate_);
```
There shouldn't be a need for adding the c++ custom ops (`TfLiteInterpreterOptionsAddCustomOp`) which was the concerning c++ part. I guess the issue is down to the c API needing the extra C++ header that you don't have installed (as original issue)?

**Comment by llschloesser** [comment]
@Namburger , that's correct, there was no need for `TfLiteInterpreterOptionsAddCustomOp`. Yes, issue is still valid, edgetpu_c.h should just point straight to the C API header.

**Comment by Namburger** [comment]
Sure, a feature request is now in place. Thanks for reporting.
